src-self-hosted/dep_tokenizer.zig

```diff
@@ -0,0 +1,1131 @@
+                const bytes = prereq.toSlice();
+                self.state = State{ .lhs = {} };
+                return Token{ .id = .prereq, .bytes = bytes };
+            },
+            .prereq_continuation_linefeed => |prereq| {
+                const bytes = prereq.toSlice();
+                self.state = State{ .lhs = {} };
+                return Token{ .id = .prereq, .bytes = bytes };
+            },
+        }
+        return null;
+    }
+
+    fn errorf(self: *Tokenizer, comptime fmt: []const u8, args: ...) Error {
+        self.error_text = (try std.Buffer.allocPrint(&self.arena.allocator, fmt, args)).toSlice();
+        return Error.InvalidInput;
+    }
+
+    fn errorPosition(self: *Tokenizer, position: usize, bytes: []const u8, comptime fmt: []const u8, args: ...) Error {
+        var buffer = try std.Buffer.initSize(&self.arena.allocator, 0);
+        std.fmt.format(&buffer, anyerror, std.Buffer.append, fmt, args) catch {};
+        try buffer.append(" '");
+        var out = makeOutput(std.Buffer.append, &buffer);
+        try printCharValues(&out, bytes);
+        try buffer.append("'");
+        std.fmt.format(&buffer, anyerror, std.Buffer.append, " at position {}", position - (bytes.len - 1)) catch {};
+        self.error_text = buffer.toSlice();
+        return Error.InvalidInput;
+    }
+
+    fn errorIllegalChar(self: *Tokenizer, position: usize, char: u8, comptime fmt: []const u8, args: ...) Error {
+        var buffer = try std.Buffer.initSize(&self.arena.allocator, 0);
+        try buffer.append("illegal char ");
+        var out = makeOutput(std.Buffer.append, &buffer);
+        try printUnderstandableChar(&out, char);
+        std.fmt.format(&buffer, anyerror, std.Buffer.append, " at position {}", position) catch {};
+        if (fmt.len != 0) std.fmt.format(&buffer, anyerror, std.Buffer.append, ": " ++ fmt, args) catch {};
+        self.error_text = buffer.toSlice();
+        return Error.InvalidInput;
+    }
+
+    const Error = error{
+        OutOfMemory,
+        InvalidInput,
+    };
+
+    const State = union(enum) {
+        lhs: void,
+        target: std.Buffer,
+        target_reverse_solidus: std.Buffer,
+        target_dollar_sign: std.Buffer,
+        target_colon: std.Buffer,
+        target_colon_reverse_solidus: std.Buffer,
+        rhs: void,
+        rhs_continuation: void,
+        rhs_continuation_linefeed: void,
+        prereq_quote: std.Buffer,
+        prereq: std.Buffer,
+        prereq_continuation: std.Buffer,
+        prereq_continuation_linefeed: std.Buffer,
+    };
+
+    const Token = struct {
+        id: ID,
+        bytes: []const u8,
+
+        const ID = enum {
+            target,
+            prereq,
+        };
+    };
+};
+
+// stage1 compiler support
+var stage2_da = std.heap.DirectAllocator.init();
+
+export fn stage2_DepTokenizer_init(input: [*]const u8, len: usize) stage2_DepTokenizer {
+    const t = stage2_da.allocator.create(Tokenizer) catch unreachable;
```

Comment:
Please don't invoke undefined behavior when out of memory occurs. You can panic - that's what stage1 currently does on OOM.